### PR TITLE
io-299/ Finance update when changing dates in project card

### PR DESCRIPTION
**Pull request by Git-Uzair**
- Added logic to move finances correctly when changing dates in project card

### CODE REVIEW

File: src/utils/common.ts [hunk 1]
```diff
@@ -129,10 +137,133 @@ export const dirtyFieldsToRequestObject = (dirtyFields: object, form: IAppForms)
...
+    budgetObject: IProjectFinances,
+    projectTimelineDates: IProjectEstDates,
+  ): IProjectFinances => {
+    /**
+     * Moves the provided finance value to its new finance location and sets current to null
+     * @param newFieldYear year for the new finance field which needs to be updated
+     * @param financeValue value which will be summed in the new year
+     * @param currentFinanceFieldName finance field name which the value belongs to
+     */
+    const updateFinanceValue = (
```
Comment (kgunga) [inline src/utils/common.ts:165]: Move this function outside the adjustBudgets function and provider parameter for adjustedBudgetObjet. Currently it is very confusing implementation

File: src/utils/common.ts [hunk 2]
```diff
@@ -129,10 +137,133 @@ export const dirtyFieldsToRequestObject = (dirtyFields: object, form: IAppForms)
...
+    const constructionEndYear = estConstructionEnd ? getYear(estConstructionEnd) : null;
+
+    const adjustedBudgetObject: IProjectFinances = {
+      ...budgetObject,
+    };
+    // looping over all budgetfields in the project
+    let fieldName: keyof IProjectFinances;
+    for (fieldName in adjustedBudgetObject) {
+      if (fieldName !== 'year') {
+        // current budget field year, extract from field name
```
Comment (kgunga) [inline src/utils/common.ts:204]: current budget field year is year itself in adjustedBudgetObject. Here logic tries to read years in future. Write a better comment here, please

**Review by kgunga — CHANGES_REQUESTED** [review]

File: src/components/Project/ProjectBasics/ProjectForm/ProjectForm.tsx [hunk 3]
```diff
@@ -66,8 +66,12 @@ const ProjectForm = () => {
...
         dispatch(setIsSaving(true));
 
-        const data: IProjectRequest = dirtyFieldsToRequestObject(dirtyFields, form as IAppForms);
-
+        const data: IProjectRequest = dirtyFieldsToRequestObject(
+          dirtyFields,
+          form as IAppForms,
+          project,
+        );
+        // syncFinancesIfNeeded(project, data);
```
Comment (kgunga) [inline src/components/Project/ProjectBasics/ProjectForm/ProjectForm.tsx:74]: Remove commented line or is this forgotten to be uncommented?

File: src/utils/common.ts [hunk 4]
```diff
@@ -129,10 +137,137 @@ export const dirtyFieldsToRequestObject = (dirtyFields: object, form: IAppForms)
...
+      8: 'preliminaryCurrentYearPlus8',
+      9: 'preliminaryCurrentYearPlus9',
+      10: 'preliminaryCurrentYearPlus10',
+    };
+    const newFinanceFieldName: keyof IProjectFinances =
+      yearToBudgetKey[newFieldYear - budgetObject.year];
+    (budgetObject[newFinanceFieldName] as string) = String(
+      safeParseInt(budgetObject[newFinanceFieldName]) + financeValue,
+    );
+    (budgetObject[currentFinanceFieldName] as null) = null;
```
Comment (kgunga) [inline src/utils/common.ts:187]: Why budgetObject[currentFinanceFieldName] casted as null and set its value to null? Is intension to delete the value or just set to null?

**Review by kgunga — CHANGES_REQUESTED** [review]

File: src/utils/common.ts (hunk 4 above)
Comment (Git-Uzair) [reply src/utils/common.ts:187]: To set it to null, the field has to be casted as null or else it throws an error. 
Intention is to delete the finance value from that field, it is set to null to make sure it is not visible in planning view timeline.

**Review by kgunga — APPROVED** [review]
👍

**Pull request merged**